src/sys/signal.rs

```diff
@@ -831,6 +831,127 @@ mod sigevent {
     }
 }
 
+/// A wrapper for `stack_t`, used to represent alternate stacks for `sigaltstack`.
+///
+/// See [`sigaltstack`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/sigaltstack.html)
+/// details on the requirements for an alternate signal stack.
+#[derive(Copy, Clone)]
+#[allow(missing_debug_implementations)]
+pub struct SigStack {
+    stack: libc::stack_t,
+}
+
+impl SigStack {
+    /// Define an alternate signal stack for use with `sigaction` using a raw pointer.
+    ///
+    /// The memory pointed to by `sp` must remain valid for as long as any signal handlers are
+    /// registered to use this stack.
+    ///
+    /// ```
+    /// # use nix::libc;
+    /// # use nix::sys::mman::{mmap, munmap, MapFlags, ProtFlags};
+    /// # use nix::sys::signal::{sigaltstack, SigStack, SigStackFlags};
+    /// # use std::ptr;
+    /// let stack_mem = unsafe {
+    ///     mmap(
+    ///         ptr::null_mut(),
+    ///         libc::SIGSTKSZ,
+    ///         ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+    ///         MapFlags::MAP_ANON | MapFlags::MAP_PRIVATE,
+    ///         -1,
+    ///         0,
+    ///     )
+    ///     .unwrap()
+    /// };
+    /// let sigstack = SigStack::from_raw(stack_mem, SigStackFlags::empty(), libc::SIGSTKSZ);
+    /// unsafe {
+    ///     let old_sigstack = sigaltstack(&sigstack).unwrap();
+    ///     sigaltstack(&old_sigstack).unwrap();
+    ///     // unmap memory only after the stack is unregistered
+    ///     munmap(stack_mem, libc::SIGSTKSZ);
+    /// }
+    /// ```
+    pub fn from_raw(sp: *mut libc::c_void, flags: SigStackFlags, size: libc::size_t) -> SigStack {
+        let mut st = unsafe { mem::uninitialized::<libc::stack_t>() };
+        st.ss_sp = sp;
+        st.ss_flags = flags.bits();
+        st.ss_size = size;
+        SigStack { stack: st }
+    }
+
+    /// Define an alternate signal stack for use with `sigaction` using a mutable byte slice.
+    ///
+    /// The `stack` slice must live at least as long as any signal handlers are registered to use
+    /// this stack. Since the lifetime of `stack` is not reflected in the type of `SigStack`, this
```

Comment:
You can ensure that the lifetime of `stack` contains the lifetime of the slice by adding a `std::marker::PhantomData` field.  In fact, do you think you should also add a `Drop` implementation that deregisters the signal stack?  That would ensure that the byte slice never gets dropped while it's still being used.

Reply:
Unfortunately the lifetime story is more complicated than that; calling `sigaltstack` essentially writes the stack pointer to a thread-local variable out in FFI land. We could add a lifetime parameter to `SigStack`, but we can't thread that lifetime through FFI land, so that parameter would be bogus in the return type of `sigaltstack`.

We also can't unregister in the `Drop` impl, because if other code has come along and also called `sigaltstack`, we might end up clobbering someone else's stack.